File: frontend/src/stores/license.ts
/**
 * Store para gestion de licencias
 */
import { defineStore } from 'pinia'
import { api } from '@/services/apiClient'
import { ref, computed } from 'vue'

// Tipos de licencia
export type LicenseTier = 'corrector' | 'profesional' | 'editorial'
export type LicenseFeature =
  | 'attribute_consistency'
  | 'grammar_spelling'
  | 'ner_coreference'
  | 'name_variants'
  | 'character_profiling'
  | 'network_analysis'
  | 'anachronism_detection'
  | 'ooc_detection'
  | 'classical_spanish'
  | 'multi_model'
  | 'full_reports'
export type LicenseStatus = 'no_license' | 'active' | 'expired' | 'grace_period' | 'suspended'
export type QuotaWarningLevel = 'none' | 'warning' | 'danger' | 'critical'

export interface LicenseInfo {
  status: LicenseStatus
  tier: LicenseTier | null
  features: LicenseFeature[]
  devices_used: number
  devices_max: number
  pages_used: number
  pages_max: number
  pages_remaining: number | null
  unlimited: boolean
  expires_at: string | null
  is_trial: boolean
  is_founding_member?: boolean
  founding_discount_eur?: number
  is_comp?: boolean
  comp_type?: string | null
  offline_days_remaining: number | null
}

export interface DeviceInfo {
  id: string
  fingerprint: string
  name: string
  status: 'active' | 'inactive' | 'pending'
  last_seen: string | null
  is_current: boolean
}

export interface UsageInfo {
  pages_used: number
  pages_max: number
  pages_remaining: number
  unlimited: boolean
}

export interface QuotaStatus {
  pages_used: number
  pages_max: number
  pages_remaining: number
  percentage: number
  warning_level: QuotaWarningLevel
  days_remaining_in_period: number
  unlimited: boolean
}



export const useLicenseStore = defineStore('license', () => {
  // Estado
  const loading = ref(false)
  const error = ref<string | null>(null)
  const licenseInfo = ref<LicenseInfo | null>(null)
  const devices = ref<DeviceInfo[]>([])
  const usage = ref<UsageInfo | null>(null)
  const quotaStatus = ref<QuotaStatus | null>(null)

  // Computados
  const isLicensed = computed(() => {
    return licenseInfo.value?.status === 'active' || licenseInfo.value?.status === 'grace_period'
  })

  const isTrial = computed(() => licenseInfo.value?.is_trial ?? false)

  const isOffline = computed(() => {
    return licenseInfo.value?.status === 'grace_period'
  })

  const offlineDaysRemaining = computed(() => {
    return licenseInfo.value?.offline_days_remaining ?? null
  })

  const tier = computed(() => licenseInfo.value?.tier ?? null)

  const tierDisplayName = computed(() => {
    const names: Record<LicenseTier, string> = {
      corrector: 'Corrector',
      profesional: 'Profesional',
      editorial: 'Editorial',
    }
    return tier.value ? names[tier.value] : 'Sin licencia'
  })

  const features = computed(() => licenseInfo.value?.features ?? [])

  const hasFeature = (feature: LicenseFeature) => {
    return features.value.includes(feature)
  }

  const pagesRemaining = computed(() => {
    if (!licenseInfo.value) return 0
    if (licenseInfo.value.unlimited) return -1 // Ilimitado
    return licenseInfo.value.pages_remaining ?? 0
  })

  const devicesRemaining = computed(() => {
    if (!licenseInfo.value) return 0
    return Math.max(0, licenseInfo.value.devices_max - licenseInfo.value.devices_used)
  })

  const quotaPercentage = computed(() => {
    if (!licenseInfo.value || licenseInfo.value.unlimited || licenseInfo.value.pages_max <= 0) return 0
    return Math.round((licenseInfo.value.pages_used / licenseInfo.value.pages_max) * 100)
  })

  const quotaWarningLevel = computed<QuotaWarningLevel>(() => {
    const pct = quotaPercentage.value
    if (pct >= 100) return 'critical'
    if (pct >= 90) return 'danger'
    if (pct >= 80) return 'warning'
    return 'none'
  })

  // Acciones
  async function fetchLicenseStatus() {
    loading.value = true
    error.value = null

    try {
      licenseInfo.value = await api.get<LicenseInfo>('/api/license/status')
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'No se pudo verificar la licencia'
      console.error('Error fetching license status:', e)
    } finally {
      loading.value = false
    }
  }

  async function activateLicense(licenseKey: string): Promise<boolean> {
    loading.value = true
    error.value = null

    try {
      await api.post('/api/license/activate', { license_key: licenseKey })
      await fetchLicenseStatus()
      return true
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Error activando licencia'
      console.error('Error activating license:', e)
      return false
    } finally {
      loading.value = false
    }
  }

  async function verifyLicense(): Promise<boolean> {
    loading.value = true
    error.value = null

    try {
      const data = await api.post<{ valid: boolean; message?: string }>('/api/license/verify')
      if (data.valid) {
        await fetchLicenseStatus()
        return true
      }
      error.value = data.message || 'Verificación fallida'
      return false
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Verificación fallida'
      console.error('Error verifying license:', e)
      return false
    } finally {
      loading.value = false
    }
  }

  async function fetchDevices() {
    try {
      const data = await api.get<{ devices: DeviceInfo[] }>('/api/license/devices')
      devices.value = data.devices || []
    } catch (e) {
      console.error('Error fetching devices:', e)
    }
  }

  async function deactivateDevice(fingerprint: string): Promise<boolean> {
    loading.value = true
    error.value = null

    try {
      await api.post('/api/license/devices/deactivate', { device_fingerprint: fingerprint })
      await fetchDevices()
      return true
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Error desactivando dispositivo'
      console.error('Error deactivating device:', e)
      return false
    } finally {
      loading.value = false
    }
  }

  async function fetchUsage() {
    try {
      usage.value = await api.get<UsageInfo>('/api/license/usage')
    } catch (e) {
      console.error('Error fetching usage:', e)
    }
  }

  async function fetchQuotaStatus() {
    try {
      quotaStatus.value = await api.get<QuotaStatus>('/api/license/quota-status')
    } catch (e) {
      console.error('Error fetching quota status:', e)
    }
  }

  const daysRemainingInPeriod = computed(() => quotaStatus.value?.days_remaining_in_period ?? null)

  async function checkFeatureAccess(featureName: LicenseFeature): Promise<boolean> {
    try {
      const data = await api.get<{ has_access: boolean }>(`/api/license/check-feature/${featureName}`)
      return data.has_access === true
    } catch (e) {
      console.error('Error checking feature access:', e)
      return false
    }
  }

  async function recordUsage(projectId: number): Promise<boolean> {
    try {
      const data = await api.post<{ allowed: boolean }>('/api/license/record-usage', { project_id: projectId })
      if (data.allowed) {
        await fetchLicenseStatus()
        return true
      }
      error.value = 'Cuota de paginas excedida'
      return false
    } catch (e) {
      console.error('Error recording usage:', e)
      return true // Permitir en modo desarrollo si no hay conexion
    }
  }

  // Inicializacion
  fetchLicenseStatus()
  fetchQuotaStatus()

  return {
    // Estado
    loading,
    error,
    licenseInfo,
    devices,
    usage,
    quotaStatus,

    // Computados
    isLicensed,
    isTrial,
    isOffline,
    offlineDaysRemaining,
    tier,
    tierDisplayName,
    features,
    pagesRemaining,
    devicesRemaining,
    quotaPercentage,
    quotaWarningLevel,
    daysRemainingInPeriod,

    // Acciones
    fetchLicenseStatus,
    activateLicense,
    verifyLicense,
    fetchDevices,
    deactivateDevice,
    fetchUsage,
    fetchQuotaStatus,
    checkFeatureAccess,
    recordUsage,
    hasFeature,
  }
})

File: api-server/routers/license.py
"""
Router: license

Endpoints para gestion de licencias, dispositivos, cuotas (paginas) y features por tier.
"""

from datetime import datetime

from deps import ApiResponse, DeviceDeactivationRequest, LicenseActivationRequest, logger
from fastapi import APIRouter, Body

router = APIRouter()


def get_license_verifier():
    """Obtiene el LicenseVerifier, retorna None si el modulo de licencias no esta disponible."""
    try:
        from narrative_assistant.licensing.verification import LicenseVerifier
        return LicenseVerifier()
    except Exception:
        return None


@router.get("/api/license/status", response_model=ApiResponse)
def get_license_status():
    """
    Obtiene el estado actual de la licencia.

    Returns:
        ApiResponse con informacion de la licencia activa
    """
    try:
        verifier = get_license_verifier()
        if not verifier:
            return ApiResponse(
                success=True,
                data={
                    "status": "no_license",
                    "tier": None,
                    "features": [],
                    "devices_used": 0,
                    "devices_max": 0,
                    "pages_used": 0,
                    "pages_max": 0,
                    "pages_remaining": 0,
                    "expires_at": None,
                    "is_trial": False,
                    "is_founding_member": False,
                    "offline_days_remaining": None,
                }
            )

        result = verifier.verify()

        if result.is_failure:
            return ApiResponse(
                success=True,
                data={
                    "status": "no_license",
                    "error": str(result.error),
                    "tier": None,
                    "features": [],
                }
            )

        verification = result.value
        license_obj = verification.license

        if license_obj is None:
            return ApiResponse(
                success=True,
                data={
                    "status": "no_license",
                    "tier": None,
                    "features": [],
                }
            )

        return ApiResponse(
            success=True,
            data={
                "status": license_obj.status.value,
                "tier": license_obj.tier.value,
                "features": [f.value for f in sorted(license_obj.features, key=lambda x: x.value)],
                "devices_used": license_obj.active_device_count,
                "devices_max": license_obj.limits.max_devices,
                "pages_used": (
                    license_obj.limits.max_pages_per_month - verification.quota_remaining
                    if verification.quota_remaining is not None
                    else 0
                ),
                "pages_max": license_obj.limits.max_pages_per_month,
                "pages_remaining": verification.quota_remaining,
                "unlimited": license_obj.limits.is_unlimited,
                "expires_at": license_obj.expires_at.isoformat() if license_obj.expires_at else None,
                "is_trial": (
                    license_obj.subscription.status == "trialing"
                    if license_obj.subscription
                    else False
                ),
                "is_founding_member": license_obj.is_founding_member,
                "founding_discount_eur": license_obj.founding_discount_eur,
                "is_comp": license_obj.is_comp,
                "comp_type": license_obj.comp_type.value if license_obj.comp_type else None,
                "offline_days_remaining": (
                    license_obj.grace_period_remaining.days
                    if license_obj.grace_period_remaining
                    else None
                ),
            }
        )

    except Exception as e:
        logger.error(f"Error getting license status: {e}", exc_info=True)
        return ApiResponse(success=False, error="Error interno del servidor")


@router.post("/api/license/activate", response_model=ApiResponse)
def activate_license(request: LicenseActivationRequest):
    """
    Activa una licencia con la clave proporcionada.

    Args:
        request: Clave de licencia

    Returns:
        ApiResponse con resultado de la activacion
    """
    try:
        verifier = get_license_verifier()
        if not verifier:
            return ApiResponse(success=False, error="Modulo de licencias no disponible")

        result = verifier.activate_license(request.license_key)

        if result.is_failure:
            return ApiResponse(success=False, error=str(result.error))

        license_obj = result.value
        return ApiResponse(
            success=True,
            data={
                "message": "Licencia activada correctamente",
                "tier": license_obj.tier.value,
                "features": [f.value for f in sorted(license_obj.features, key=lambda x: x.value)],
            }
        )

    except Exception as e:
        logger.error(f"Error activating license: {e}", exc_info=True)
        return ApiResponse(success=False, error="Error interno del servidor")


@router.post("/api/license/verify", response_model=ApiResponse)
def verify_license():
    """
    Verifica la licencia actual (online si es posible).

    Returns:
        ApiResponse con resultado de la verificacion
    """
    try:
        verifier = get_license_verifier()
        if not verifier:
            return ApiResponse(success=False, error="Modulo de licencias no disponible")

        result = verifier.verify(force_online=True)

        if result.is_failure:
            return ApiResponse(success=False, error=str(result.error))

        return ApiResponse(
            success=True,
            data={
                "valid": result.value.is_valid,
                "message": result.value.message,
            }
        )

    except Exception as e:
        logger.error(f"Error verifying license: {e}", exc_info=True)
        return ApiResponse(success=False, error="Error interno del servidor")


@router.get("/api/license/devices", response_model=ApiResponse)
def get_license_devices():
    """
    Obtiene la lista de dispositivos registrados en la licencia.

    Returns:
        ApiResponse con lista de dispositivos
    """
    try:
        verifier = get_license_verifier()
        if not verifier:
            return ApiResponse(success=False, error="Modulo de licencias no disponible")

        result = verifier.verify()

        if result.is_failure:
            return ApiResponse(success=False, error=str(result.error))

        license_obj = result.value.license
        if not license_obj:
            return ApiResponse(success=False, error="No se encontro licencia")

        return ApiResponse(
            success=True,
            data={
                "devices": [
                    {
                        "id": d.id,
                        "fingerprint": d.hardware_fingerprint[:8] + "...",
                        "name": d.device_name,
                        "status": d.status.value,
                        "last_seen": d.last_seen_at.isoformat() if d.last_seen_at else None,
                        "is_current": d.is_current_device,
                    }
                    for d in license_obj.devices
                ],
                "max_devices": license_obj.limits.max_devices,
            }
        )

    except Exception as e:
        logger.error(f"Error getting devices: {e}", exc_info=True)
        return ApiResponse(success=False, error="Error interno del servidor")


@router.post("/api/license/devices/deactivate", response_model=ApiResponse)
def deactivate_device(request: DeviceDeactivationRequest):
    """
    Desactiva un dispositivo de la licencia.

    Args:
        request: Fingerprint del dispositivo a desactivar

    Returns:
        ApiResponse con resultado de la desactivacion
    """
    try:
        verifier = get_license_verifier()
        if not verifier:
            return ApiResponse(success=False, error="Modulo de licencias no disponible")

        result = verifier.deactivate_device(request.device_fingerprint)

        if result.is_failure:
            return ApiResponse(success=False, error=str(result.error))

        return ApiResponse(
            success=True,
            data={
                "message": "Dispositivo desactivado",
                "cooldown_hours": 48,
            }
        )

    except Exception as e:
        logger.error(f"Error deactivating device: {e}", exc_info=True)
        return ApiResponse(success=False, error="Error interno del servidor")


@router.get("/api/license/usage", response_model=ApiResponse)
def get_license_usage():
    """
    Obtiene el uso de la licencia en el periodo actual (paginas).

    Returns:
        ApiResponse con estadisticas de uso en paginas
    """
    try:
        verifier = get_license_verifier()
        if not verifier:
            return ApiResponse(success=False, error="Modulo de licencias no disponible")

        result = verifier.verify()

        if result.is_failure:
            return ApiResponse(success=False, error=str(result.error))

        verification = result.value
        license_obj = verification.license

        if not license_obj:
            return ApiResponse(success=False, error="No se encontro licencia")

        limits = license_obj.limits
        quota_remaining = verification.quota_remaining

        return ApiResponse(
            success=True,
            data={
                "pages_used": (
                    limits.max_pages_per_month - quota_remaining
                    if quota_remaining is not None
                    else 0
                ),
                "pages_max": limits.max_pages_per_month,
                "pages_remaining": quota_remaining if quota_remaining is not None else -1,
                "unlimited": limits.is_unlimited,
            }
        )

    except Exception as e:
        logger.error(f"Error getting usage: {e}", exc_info=True)
        return ApiResponse(success=False, error="Error interno del servidor")


@router.get("/api/license/quota-status", response_model=ApiResponse)
def get_quota_status():
    """
    Estado ligero de cuota con porcentaje, nivel de aviso y dias restantes del periodo.

    Returns:
        ApiResponse con pages_used, pages_max, pages_remaining, percentage,
        warning_level, days_remaining_in_period, unlimited.
    """
    try:
        verifier = get_license_verifier()
        if not verifier:
            return ApiResponse(
                success=True,
                data={
                    "pages_used": 0,
                    "pages_max": 0,
                    "pages_remaining": 0,
                    "percentage": 0,
                    "warning_level": "none",
                    "days_remaining_in_period": 0,
                    "unlimited": False,
                },
            )

        result = verifier.verify()
        if result.is_failure:
            return ApiResponse(
                success=True,
                data={
                    "pages_used": 0,
                    "pages_max": 0,
                    "pages_remaining": 0,
                    "percentage": 0,
                    "warning_level": "none",
                    "days_remaining_in_period": 0,
                    "unlimited": False,
                },
            )

        verification = result.value
        license_obj = verification.license

        if not license_obj:
            return ApiResponse(
                success=True,
                data={
                    "pages_used": 0,
                    "pages_max": 0,
                    "pages_remaining": 0,
                    "percentage": 0,
                    "warning_level": "none",
                    "days_remaining_in_period": 0,
                    "unlimited": False,
                },
            )

        limits = license_obj.limits
        is_unlimited = limits.is_unlimited

        # Calcular paginas
        quota_remaining = verification.quota_remaining
        pages_max = limits.max_pages_per_month
        pages_used = (
            pages_max - quota_remaining
            if quota_remaining is not None
            else 0
        )
        pages_remaining = quota_remaining if quota_remaining is not None else 0

        # Porcentaje y nivel de aviso
        if is_unlimited or pages_max <= 0:
            percentage = 0
            warning_level = "none"
        else:
            percentage = min(100, round(pages_used / pages_max * 100))
            if percentage >= 100:
                warning_level = "critical"
            elif percentage >= 90:
                warning_level = "danger"
            elif percentage >= 80:
                warning_level = "warning"
            else:
                warning_level = "none"

        # Dias restantes en el periodo de facturacion
        days_remaining_in_period = 0
        if license_obj.subscription and license_obj.subscription.current_period_end:
            delta = license_obj.subscription.current_period_end - datetime.utcnow()
            days_remaining_in_period = max(0, delta.days)
        else:
            # Fallback: estimar fin de mes natural
            now = datetime.utcnow()
            if now.month == 12:
                next_month = now.replace(year=now.year + 1, month=1, day=1)
            else:
                next_month = now.replace(month=now.month + 1, day=1)
            days_remaining_in_period = max(0, (next_month - now).days)

        return ApiResponse(
            success=True,
            data={
                "pages_used": pages_used,
                "pages_max": pages_max,
                "pages_remaining": pages_remaining,
                "percentage": percentage,
                "warning_level": warning_level,
                "days_remaining_in_period": days_remaining_in_period,
                "unlimited": is_unlimited,
            },
        )

    except Exception as e:
        logger.error(f"Error getting quota status: {e}", exc_info=True)
        return ApiResponse(success=False, error="Error interno del servidor")


@router.post("/api/license/record-usage", response_model=ApiResponse)
def record_usage(project_id: int = Body(..., embed=True)):
    """
    Registra el uso de un manuscrito contra la cuota de paginas.

    Args:
        project_id: ID del proyecto/manuscrito

    Returns:
        ApiResponse con resultado del registro
    """
    try:
        verifier = get_license_verifier()
        if not verifier:
            # Sin verificador, permitir uso (desarrollo)
            return ApiResponse(success=True, data={"allowed": True})

        result = verifier.check_quota()

        if result.is_failure:
            error = result.error
            return ApiResponse(
                success=False,
                error=str(error),
                data={"allowed": False, "reason": error.__class__.__name__}
            )

        return ApiResponse(
            success=True,
            data={
                "allowed": True,
                "pages_remaining": result.value,
            }
        )

    except Exception as e:
        logger.error(f"Error recording usage: {e}", exc_info=True)
        return ApiResponse(success=False, error="Error interno del servidor")


@router.get("/api/license/check-feature/{feature_name}", response_model=ApiResponse)
def check_feature_access(feature_name: str):
    """
    Verifica si el usuario tiene acceso a una feature segun su tier.

    Args:
        feature_name: Nombre de la feature (attribute_consistency, character_profiling, etc.)

    Returns:
        ApiResponse indicando si tiene acceso
    """
    try:
        verifier = get_license_verifier()
        if not verifier:
            # Sin verificador, permitir todo (desarrollo)
            return ApiResponse(success=True, data={"has_access": True})

        from narrative_assistant.licensing.models import LicenseFeature

        try:
            feature = LicenseFeature(feature_name)
        except ValueError:
            return ApiResponse(
                success=False,
                error=f"Feature desconocida: {feature_name}",
            )

        result = verifier.check_feature(feature)

        if result.is_failure:
            return ApiResponse(
                success=True,
                data={
                    "has_access": False,
                    "reason": str(result.error),
                    "required_tier": "profesional",
                }
            )

        return ApiResponse(
            success=True,
            data={
                "has_access": True,
                "feature": feature_name,
            }
        )

    except Exception as e:
        logger.error(f"Error checking feature access: {e}", exc_info=True)
        return ApiResponse(success=False, error="Error interno del servidor")
